### DISCUSSION TITLE
Config and Custom Costs

### DISCUSSION THREAD

**Discussion by Kashian** [valhalla/valhalla#4254] [Q&A] [answered]
I have gone through almost all discussion posts looking for custom costs and easy ways of implementing such dynamic costing.
I have seen SIF component in the Valhalla project. However it seems I have to touch C++ codes, and compile/build the project myself. As a six sense, I was telling myself, hey man, we are in 2023, so there must be easier way to assign dynamic costs to edges and nodes. I have seen examples on how to set filters and use some options in the JSON request to force the system to apply such filters in calculating costs, including penalties, etc. However I am not sure if we can simply create a profile in the JSON config file or even for example, if we can include following lines inside the config file:
```
"costing": "truck",
            "costing_options": {
                "truck": {"gate_penalty": 600, "length": 4, "weight": 16, "width": 2, "height": 5}
            }

```
The above code is taken from #3295 
I am not sure if based on above example, we can add costing_options via valhalla.json file.
Also I need to assign dynamic cost to specific road classes based on some criteria that we have. The criteria is complex. For example we do not want to pass through roads which are considered as residential and the speed limit is below 20 and the road name is A or B or C in its name. This is just an example. Or we would like to only do Turn Left if a traffic control device exists (A.K.A Protected Left Turn).
I appreciate if you can point me toward some sample scripts which implements costing_options or sample Config file which has these options. My default config file has no Costing keyword.
Thanks

**Comment by nilsnolde** [comment]
Yeah, if you want to implement _new_ costing options, you'll have to touch C++, no way around it. 2023 or 1999, year is irrelevant. There's many PRs implementing new costing options, you'll have to search a bit for examples.

To _use_ costing options, you do that via request parameters, not a  (static) config. For such specific things (avoid residential below 20 etc), there's no interface in Valhalla. If you know the locations of certain roads, you can use `exclude_locations` parameter.

**Reply by Kashian** [reply]
Thanks for your comment.
To elaborate the scenario, I created a sample sketch which shows two scenarios:

![Scenario-SharpTurns](https://github.com/valhalla/valhalla/assets/5819931/6ac85989-b9f1-460f-82e5-bcb26e18facd)

This 1st Vehicle and 2nd Vehicle are both large Trailer Trucks. 
The 1st Truck needs to do a sharp turn to right, however due to the road width constraints or road curvature limits, the truck can't complete the maneuver. 
The 2nd Truck needs to turn left, but only allowed if Traffic light exists so it can complete a protected Left Turn. In this case, processing existence of control devices at each intersection is important and should be taken as a cost factor.
For example, in case of no traffic lights, we avoid left turn (high cost).

I appreciate your comment about above scenario and if OSM dataset can't help about such limits, what would be our options to go forward?

You mentioned about some Pull Requests, creating custom costing options. Can you link me to two samples?

**Reply by kevinkreiser** [reply]
excellent illustration!

vehicle 1: we have an approximation of the tightness of the turn and could bar trucks from doing turns that are too tight even as a costing parameter. adding such a thing would be interesting but also could make for some crazy looking routes, not saying its a bad idea though just could be intersting.
vehicle 2: osm data is not well populated enough for us to accurately penalize this. we could parse this information where it exists in osm, but if we penalize lefts which dont have left turn signals they will be so sparse the router will just take the next one which doesnt happen to be properly marked in osm. now you could say commercial data like tomtom or HERE has this data potentially and support it that way which is ok but we need a way at run time to turn on and off the penalty so that you can decide if the penalty makes sense to use based on the density and accuracy of the tagging in the data.

**Reply by kevinkreiser** [reply, marked as answer]
and if you are looking for PRs you should go to the costing impelmentation and click the `blame` button. then you can scroll down the left and side and see which pr titles sound like new costing or costing options. there are some good ones in there: 
![image](https://github.com/valhalla/valhalla/assets/697548/a84dc992-95b6-4ed2-853a-e4a7108036b8)
